From 899f6f0d865d238a68c20b641cf7e636a1ecf5ea Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Micha=C5=82=20G=C3=B3rny?= <mgorny@gentoo.org>
Date: Wed, 27 Nov 2024 13:47:23 +0100
Subject: [PATCH 03/22] Allow overriding CUDA-related paths

The second hunk used to extend the env-var forwarding loop in
tools/setup_helpers/cmake.py to also pass CUDA* variables through to CMake.
Upstream PR #177640 moved that forwarding loop into
cmake/EnvVarForwarding.cmake, so apply the same change there instead.
---
 cmake/EnvVarForwarding.cmake        | 8 ++++++++
 cmake/Modules/FindCUDAToolkit.cmake | 2 +-
 2 files changed, 9 insertions(+), 1 deletion(-)

diff --git a/cmake/EnvVarForwarding.cmake b/cmake/EnvVarForwarding.cmake
index a31ce6db266..86526892cb3 100644
--- a/cmake/EnvVarForwarding.cmake
+++ b/cmake/EnvVarForwarding.cmake
@@ -18,6 +18,14 @@ set(_ENV_PASSTHROUGH
   CUDA_HOST_COMPILER
   CUDA_NVCC_EXECUTABLE
   CUDA_SEPARABLE_COMPILATION
+  # NB: we can't do a blanket CUDA* below since CMake forces all envvars
+  # uppercase on Windows
+  CUDAToolkit_BIN_DIR
+  CUDAToolkit_ROOT_DIR
+  CUDA_TOOLKIT_ROOT_DIR
+  CUDAToolkit_ROOT
+  CUDAToolkit_TARGET_DIR
+  CUDA_VERSION
   CUDNN_LIBRARY
   CUDNN_INCLUDE_DIR
   CUDNN_ROOT
diff --git a/cmake/Modules/FindCUDAToolkit.cmake b/cmake/Modules/FindCUDAToolkit.cmake
index ec9ae530aa6..b7c0bd9fc51 100644
--- a/cmake/Modules/FindCUDAToolkit.cmake
+++ b/cmake/Modules/FindCUDAToolkit.cmake
@@ -497,7 +497,7 @@ Result variables
 # CMAKE_CUDA_COMPILER_TOOLKIT_ROOT and CMAKE_CUDA_COMPILER_LIBRARY_ROOT.
 # We compute the rest based on those here to avoid re-searching and to avoid finding a possibly
 # different installation.
-if(CMAKE_CUDA_COMPILER_TOOLKIT_ROOT)
+if(CMAKE_CUDA_COMPILER_TOOLKIT_ROOT AND NOT "$ENV{IN_PYTORCH_BUILD}" STREQUAL "1")
   set(CUDAToolkit_ROOT_DIR "${CMAKE_CUDA_COMPILER_TOOLKIT_ROOT}")
   set(CUDAToolkit_LIBRARY_ROOT "${CMAKE_CUDA_COMPILER_LIBRARY_ROOT}")
   set(CUDAToolkit_VERSION "${CMAKE_CUDA_COMPILER_TOOLKIT_VERSION}")
